### PR TITLE
Perfdata: Add support for more UoMs

### LINKED ISSUE #71: Support for more UoMs in perfdata
## Is your feature request related to a problem? Please describe.
Check plugins currently have only a limited set of UoMs (Units of measurement) available which Icinga understands.

## Describe the solution you'd like
With Icinga v2.13 support for more is added. We have to support them as well.

## Additional context
https://github.com/Icinga/icinga2/issues/7225

### PR DESCRIPTION
**Pull request by Yonas-net**
resolves #71

### CODE REVIEW

**Comment by Yonas-net** [comment]
Depends on [#4369](https://github.com/Icinga/icingaweb2/pull/4369)

File: library/Icingadb/Util/PerfData.php [hunk 1]
```diff
@@ -307,8 +451,8 @@ protected function parse()
         $parts = explode(';', $this->perfdataValue);
 
         $matches = array();
-        if (preg_match('@^(-?\d+(\.\d+)?)([a-zA-Z%°]{1,2})$@u', $parts[0], $matches)) {
-            $this->unit = strtolower($matches[3]);
+        if (preg_match('@^(-?\d+(\.\d+)?)([a-zA-Z%°]{1,3})$@u', $parts[0], $matches)) {
+            $this->unit = $matches[3];
             $this->value = self::convert($matches[1], $this->unit);
```
Comment (nilmerg) [inline library/Icingadb/Util/PerfData.php:456]: I'm not sure if values need to be converted here still. If we expect the input data as being normalized, there's no need to. Also, doesn't the formatting in method `format()` expect the normalized values from Icinga?

File: library/Icingadb/Util/PerfData.php [hunk 2]
```diff
@@ -449,15 +643,67 @@ protected function format($value)
...
+            return Format::ohms($value);
+        }
+        if ($this->isGrams()) {
+            return Format::grams($value);
+        }
+        if ($this->isLiters()) {
+            return Format::liters($value);
+        }
         if ($this->isTemperature()) {
             return (string)$value . strtoupper($this->unit);
```
Comment (nilmerg) [inline library/Icingadb/Util/PerfData.php:704]: The `strtoupper` call isn't necessary anymore.

File: library/Icingadb/Util/PerfData.php [hunk 3]
```diff
@@ -449,15 +643,67 @@ protected function format($value)
         if ($this->isPercentage()) {
```
Comment (nilmerg) [inline library/Icingadb/Util/PerfData.php:643]: * Consider to use a `switch (true) { case $this->isPercentage(): ... }` here
  * There are now so much similar control structures, it makes sense to condense them
* Also please put a space between casts and variables everywhere (i.e. `(string) $value`)

File: library/Icingadb/Util/Format.php [hunk 4]
```diff
@@ -0,0 +1,275 @@
+<?php
+
+namespace Icinga\Module\Icingadb\Util;
+
+use DateTime;
+
+class Format
```
Comment (nilmerg) [inline library/Icingadb/Util/Format.php:7]: Please name it `PerfDataFormat` and remove all unused methods. (e.g. `secondsByMonth`)

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: library/Icingadb/Util/PerfData.php (hunk 1 above)
Comment (Yonas-net) [reply library/Icingadb/Util/PerfData.php:456]: Ohh, you're right. I wonder if this is also the case for the thresholds, so are the thresholds normalized as well or does this only affect the plugin output?

File: library/Icingadb/Util/PerfData.php [hunk 5]
```diff
@@ -166,14 +165,159 @@ public function isPercentage()
...
+
+    /**
+     * Get whether this data's value is in watt seconds
+     *
+     * @return bool
+     */
+    public function isWattSeconds()
+    {
+        return in_array($this->unit, Format::getTimeWattPrefix()['WattSeconds']);
+    }
```
[the remark this answers is not in the document]
Comment (Yonas-net) [reply library/Icingadb/Util/PerfData.php:246]: As I understand it, they are different units. For example these units ``mWm, uWm`` are normalized to ``Wm`` and not to ``Wh``. This also affects the different ampere units``Ah, Am, As``. But I might be wrong.

File: library/Icingadb/Util/PerfData.php (hunk 1 above)
Comment (nilmerg) [reply library/Icingadb/Util/PerfData.php:456]: That is a question for @Al2Klimov :wink:

File: library/Icingadb/Util/PerfData.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (nilmerg) [reply library/Icingadb/Util/PerfData.php:246]: the [table](https://github.com/Icinga/icinga2/pull/8759/files#diff-b5d3095868e71d7e5d627a517bab1e8e40198c49d0536f43e36b60ede5488278R677-R679) shows that for these units the metric backends get *watt-hours*, so it should be `Wh` for them if normalized.

File: library/Icingadb/Util/PerfData.php (hunk 1 above)
Comment (Al2Klimov) [reply library/Icingadb/Util/PerfData.php:456]: Reading about _plugin output normalization_ the first time. 😉

And – regarding the thresholds:

https://github.com/Icinga/icinga2/blob/63b53bfda92930b2332035860082dbcaa5cff9e7/test/icinga-perfdata.cpp#L46-L53

Comment (Yonas-net) [reply library/Icingadb/Util/PerfData.php:456]: > Reading about _plugin output normalization_ the first time. 😉

My mistake.

File: library/Icingadb/Util/PerfDataFormat.php [hunk 6]
```diff
@@ -0,0 +1,138 @@
+<?php
+
+/* Icinga DB Web | (c) 2021 Icinga GmbH | GPLv2+ */
```
Comment (nilmerg) [inline library/Icingadb/Util/PerfDataFormat.php:3]: Please change this to `GPLv2` (Remove the plus). I know, the other PerfData files also use it, but it's wrong as well.

File: library/Icingadb/Util/PerfDataFormat.php [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+        if ($absValue < 60) {
+            return self::formatForUnits($value, self::$secondPrefix, self::$generalBase);
+        } elseif ($absValue < 3600) {
+            return sprintf('%0.2f m', $value / 60);
+        } elseif ($absValue < 86400) {
+            return sprintf('%0.2f h', $value / 3600);
+        }
+
+        // TODO: Do we need weeks, months and years?
```
Comment (nilmerg) [inline library/Icingadb/Util/PerfDataFormat.php:107]: I'd say, after many weeks and months and years: No.

Don't keep the comment, remove it :wink:

**Review by nilmerg — CHANGES_REQUESTED** [review]
Nice! Two minor changes left and it's ready!

**Review by nilmerg — DISMISSED** [review]
LGTM! Can be merged once tested with actual normalized data. (Blocked by ~~https://github.com/Icinga/icinga2/pull/8761~~ https://github.com/Icinga/icingadb/pull/308 in the meantime)

**Review by nilmerg — APPROVED** [review]

**Pull request merged**